### PR TITLE
Use golang.org/x/xerrors instead of raw `return err`

### LINKED ISSUE #178: Start using golang.org/x/xerrors instead of raw `return err`
**Is your feature request related to a problem? Please describe.**

I am wary of the way how errors are currently returned from functions in the codebase — for example, code blocks like below:

```go
if err != nil {
    log.Error(err, "...<some message>...", "foo", bar)
    return err
}
```

As much as I understand the value of structured logs, I'm especially uncomfortable with the "naked" `return err`. This isn't a good practice in Go, in part because it loses information for tracking the details of the problem at intermediate steps, when the error finally surfaces somewhere higher.

**Describe the solution you'd like**

I suggest using https://golang.org/x/xerrors package, and through its [Errorf](https://godoc.org/golang.org/x/xerrors#Errorf) function, adding helpful context information when returning an error, instead of a raw `return err`.

**Describe alternatives you've considered**

I considered also removing `log.Error` calls; but I think those can be still useful for getting information in a more structured way (JSON), better for bulk log analysis than the format used in xerrors. 

**What version of eunomia are you using?**

eunomia version: 0.0.4 / dev

**Additional context**

### PR DESCRIPTION
**Pull request by karolkrawczyk**
**Description**

Use golang.org/x/errors so errors have a better context for
easier troubleshooting.

Fixes #178 

**Type of change**

* New feature (non-breaking change which adds functionality)

**Checklist**

- [not needed] Unit tests and e2e tests updated
- [not needed] Documentation updated

### CODE REVIEW

File: pkg/controller/gitopsconfig/controller.go [hunk 1]
```diff
@@ -244,19 +246,19 @@ func (r *Reconciler) createJob(jobtype string, instance *gitopsv1alpha1.GitOpsCo
 	job, err := util.CreateJob(mergedata)
 	if err != nil {
 		log.Error(err, "unable to create job manifest from merge data", "mergedata", mergedata)
-		return reconcile.Result{}, err
+		return reconcile.Result{}, xerrors.Errorf("unable to create job manifest from merge data %q: %w", mergedata, err)
```
Comment (akavel) [inline pkg/controller/gitopsconfig/controller.go]: I believe mergedata may be quite huge, so I'm not convinced we want to try printing it all here; also, what will `%q` print in this case?

File: pkg/controller/gitopsconfig/controller.go [hunk 2]
```diff
@@ -244,19 +246,19 @@ func (r *Reconciler) createJob(jobtype string, instance *gitopsv1alpha1.GitOpsCo
...
 	}
 	err = controllerutil.SetControllerReference(instance, &job, r.scheme)
 	if err != nil {
 		log.Error(err, "unable to the owner for job", "job", job)
-		return reconcile.Result{}, err
+		return reconcile.Result{}, xerrors.Errorf("unable to the owner for job %q: %w", job, err)
```
Comment (akavel) [inline pkg/controller/gitopsconfig/controller.go]: 1. the sentence doesn't seem to gramatically make sense to me; can you try to understand or ask what the original author meant here, and improve this message in both lines above?
2. I am not sure %q will print something sensible for whole "job" object
3. In a similar vein, please review all the remaining  uses of `%q` in this PR, to make sure they actually print something sensible and useful, while also not dumping whole huge object; please let me know afterwards, I'll come back to reviewing this PR then

File: pkg/controller/gitopsconfig/controller.go [hunk 3]
```diff
@@ -282,14 +284,14 @@ func (r *Reconciler) createCronJob(instance *gitopsv1alpha1.GitOpsConfig) error
 			update = false
 		} else {
 			// Error reading the object - requeue the request.
-			return err
+			return xerrors.Errorf("cannot read kubernetes object: %w", err)
```
Comment (akavel) [inline pkg/controller/gitopsconfig/controller.go]: please try to add more context to this message; see also comment on l.174

File: pkg/controller/gitopsconfig/controller.go [hunk 4]
```diff
@@ -282,14 +284,14 @@ func (r *Reconciler) createCronJob(instance *gitopsv1alpha1.GitOpsConfig) error
...
 		}
 	}
 
 	err = controllerutil.SetControllerReference(instance, &cronjob, r.scheme)
 	if err != nil {
 		log.Error(err, "unable to the owner for cronjob", "cronjob", cronjob)
-		return err
+		return xerrors.Errorf("unable to the owner for cronjob %q: %w", cronjob, err)
```
Comment (akavel) [inline pkg/controller/gitopsconfig/controller.go]: please improve as mentioned above for similar line

File: pkg/controller/gitopsconfig/controller.go [hunk 5]
```diff
@@ -320,7 +322,7 @@ func (r *Reconciler) GetAll() (gitopsv1alpha1.GitOpsConfigList, error) {
 	err := r.client.List(context.TODO(), &client.ListOptions{}, instanceList)
 	if err != nil {
 		log.Error(err, "unable to get the list of GitOpsCionfig")
-		return *instanceList, err
+		return *instanceList, xerrors.Errorf("unable to get the list of GitOpsCionfig: %w", err)
```
Comment (akavel) [inline pkg/controller/gitopsconfig/controller.go]: please fix typo: Cionfig -> Config

File: pkg/util/util.go [hunk 6]
```diff
@@ -56,14 +57,14 @@ func InitializeTemplates(jobTemplateFileName string, cronJobTemplateFileName str
 
 	jobTemplate, err = jobTemplate.Parse(string(text))
 	if err != nil {
-		log.Error(err, "Error parsing template", "template", text)
-		return err
+		log.Error(err, "error parsing template", "template", text)
+		return xerrors.Errorf("error parsing template %q: %w", text, err)
```
Comment (akavel) [inline pkg/util/util.go]: please don't dump whole template contents into error message

File: pkg/controller/gitopsconfig/controller.go (hunk 3 above)
Comment (karolkrawczyk) [reply pkg/controller/gitopsconfig/controller.go]: Done.

File: pkg/controller/gitopsconfig/controller.go (hunk 4 above)
Comment (karolkrawczyk) [reply pkg/controller/gitopsconfig/controller.go]: Fixed, please let me know if new error message sounds better.

File: pkg/controller/gitopsconfig/controller.go (hunk 5 above)
Comment (karolkrawczyk) [reply pkg/controller/gitopsconfig/controller.go]: Sure.

File: pkg/controller/gitopsconfig/controller.go [hunk 7]
```diff
@@ -19,6 +19,7 @@ package gitopsconfig
 import (
 	"context"
 	goerrors "errors"
+	"golang.org/x/xerrors"
```
Comment (akavel) [inline pkg/controller/gitopsconfig/controller.go]: package xerrors is not in Go standard library, so it should go in the block between l.25-l.42, here and in all files below

**Comment by seanmalloy** [comment]
Looks pretty good. Please just address the final comment from @akavel https://github.com/KohlsTechnology/eunomia/pull/257#discussion_r368627912

**Comment by karolkrawczyk** [comment]
@akavel I believe I fixed import order. Please review once again when you'll have some spare time, thank you.

File: pkg/controller/gitopsconfig/controller.go [hunk 8]
```diff
@@ -391,17 +393,17 @@ func (r *Reconciler) manageDeletion(instance *gitopsv1alpha1.GitOpsConfig) (reco
...
 	// looking up all delete jobs
 	err = r.client.List(context.TODO(), &client.ListOptions{
 		Namespace:     instance.GetNamespace(),
 		LabelSelector: selector,
 	}, jobList)
 	if err != nil {
-		log.Error(err, "unable to list jobs ")
-		return reconcile.Result{}, err
+		log.Error(err, "unable to list all delete jobs")
+		return reconcile.Result{}, xerrors.Errorf("unable to list all delete jobs: %w", err)
```
Comment (akavel) [inline pkg/controller/gitopsconfig/controller.go]: on l. 180, there's an implicit "return err"; as a result, if this error happens, we won't know which instance it's related to; please either add this info in this error message, or expand l.180 to add it

Comment (akavel) [reply pkg/controller/gitopsconfig/controller.go]: after discussion, can be left as is, thanks!

**Pull request merged**